solr/core/src/java/org/apache/solr/cluster/placement/plugins/OrderedNodePlacementPlugin.java

```diff
@@ -404,20 +404,25 @@ public abstract static class WeightedNode implements Comparable<WeightedNode> {
     private final Node node;
     private final Map<String, Map<String, Set<Replica>>> replicas;
 
+    // a flattened list of all replicas, as computing from the map could be costly
+    private final Set<Replica> allReplicas;
+
     public WeightedNode(Node node) {
       this.node = node;
       this.replicas = new HashMap<>();
+      this.allReplicas = new HashSet<>();
     }
 
     public Node getNode() {
       return node;
     }
 
     public Set<Replica> getAllReplicasOnNode() {
-      return replicas.values().stream()
-          .flatMap(shard -> shard.values().stream())
-          .flatMap(Collection::stream)
-          .collect(Collectors.toSet());
+      return new HashSet<>(allReplicas);
```

Comment:
based on how this is being used currently, there's no need to wrap this in a new HashSet. The method is public, so you never know who might call it and what their assumptions might be; but you are replacing `Collectors.toSet()`, which specifically makes no guarantees about mutability/immutability, so I'd be inclined to just do `Collections.unmodifableSet(allReplicas)`, taking this opportunity to make the API more conservative, without breaking any backcompat.

Also, if the motivating concern is "many replicas", the cost of instantiating a new HashMap that we don't really need is not worth it.

Reply:
As discussed, the current fix is to use the most defensive approach which does not modify the behavior at all. The minor concern with `unmodifableSet` (which I did use at firs) was that the "view" could still mutate after this method returns and this could break certain expectation of the caller.

Otherwise I don't have any objection using `unmodifableSet`. Perhaps @HoustonPutman can share some thoughts? 😊